ggml-sycl/common.hpp

```diff
@@ -0,0 +1,437 @@
+
+static ggml_sycl_backend_gpu_mode g_ggml_sycl_backend_gpu_mode =
+    SYCL_UNSET_GPU_MODE;
+
+struct sycl_device_capabilities {
+  int cc; // compute capability
+  bool vmm; // virtual memory support
+  size_t vmm_granularity; // granularity of virtual memory
+  int device_id;
+};
+
+static sycl_device_capabilities g_device_caps[SYCL_MAX_DEVICES] = {
+    {0, false, 0, -1}};
+
+struct sycl_device_id2index {
+  int index;
+};
+
+static void* g_scratch_buffer = nullptr;
+static size_t g_scratch_size = 0; // disabled by default
+static size_t g_scratch_offset = 0;
+
+static dpct::queue_ptr g_sycl_handles[SYCL_MAX_DEVICES] = {nullptr};
+
+int get_main_device();
+
+[[noreturn]] static inline void bad_arch(const sycl::stream& stream_ct1) {
+  stream_ct1 << "ERROR: ggml-sycl was compiled without support for the "
+                "current GPU architecture.\n";
+  // __trap();
+  std::exit(1);
+
+  (void)bad_arch; // suppress unused function warning
+}
+
+/*
+device_index: device index from 0 to n (continue numbers).
+    It is used for device select/set in SYCL backend internal data structure.
+*/
+void check_allow_gpu_index(const int device_index);
+
+/*
+device_id: device ID is shown by ggml_backend_sycl_print_sycl_devices().
+    It is only used to set current working device.
+*/
+inline void check_allow_gpu_id(const int device_id) {
+  if (!g_sycl_gpu_mgr->is_allowed_gpu(device_id)) {
+    char error_buf[256];
+    snprintf(
+        error_buf,
+        sizeof(error_buf),
+        "error: cannot set device=%d, which is not allowed. Please "
+        "set GPU ID in: [%s]",
+        device_id,
+        g_sycl_gpu_mgr->gpus_list.c_str());
+    fprintf(stderr, "%s\n", error_buf);
+    throw std::invalid_argument(error_buf);
+  }
+}
+
+
+int get_current_device_id();
+
+inline dpct::err0 ggml_sycl_set_device(const int device) try {
+  int device_id = g_sycl_gpu_mgr->gpus[device];
+  check_allow_gpu_id(device_id);
+
+  int current_device_id;
+  SYCL_CHECK(CHECK_TRY_ERROR(current_device_id = get_current_device_id()));
+
+  // GGML_SYCL_DEBUG("ggml_sycl_set_device device_id=%d,
+  // current_device_id=%d\n", device, current_device);
+  if (device_id == current_device_id) {
+    return 0;
+  }
+
+  return CHECK_TRY_ERROR(dpct::select_device(device_id));
+} catch (sycl::exception const& exc) {
+  std::cerr << exc.what() << "Exception caught at file:" << __FILE__
+            << ", line:" << __LINE__ << std::endl;
+  crash();
+  std::exit(1);
+}
+
+void log_ggml_var_device(
+    const char* name,
+    float* src,
+    size_t total_elements,
+    bool src_on_device);
+
+void log_ggml_var_device_fp16(
+    const char* name,
+    sycl::half* src,
+    size_t total_elements,
+    bool src_on_device);
+
+// todo: debug for crash in some case
+void print_ggml_tensor(const char* name, struct ggml_tensor* src);
+
+static int log_file_name_idx = 0;
+void log_tensor_with_cnt(
+    const char* name,
+    struct ggml_tensor* src,
+    int stop_cnt);
+
```

Comment:
```suggestion
template <typename T>
using to_t_sycl_t = void (*)(const void *__restrict__ x, T *__restrict__ y,
                             int k, dpct::queue_ptr stream);
typedef to_t_sycl_t<float> to_fp32_sycl_t;
typedef to_t_sycl_t<sycl::half> to_fp16_sycl_t;
```
The type conversion can be added in  common.hpp  so that it can be called in dequantize and in turn in dmmv.cpp
Otherwise there is build error. Also these lines from ggml-sycl.cpp can be removed after adding here.

Reply:
If need to move these convert type function to other files, I suggest to move to convert.cpp/hpp, like CUDA.
The file name of convert.cpp has better meaning than common.cpp

Reply:
@NeoZhangJianyu fixed